mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,424 @@
+        let mut attempts = 0;
+
+        loop {
+            match tokio_tungstenite_wasm::connect(self.node.uri.to_string()).await {
+                Ok(ws) => return Ok(ws),
+                Err(e) => {
+                    attempts += 1;
+                    if attempts > max_attempts {
+                        return Err(e);
+                    }
+
+                    Timer::sleep(sleep_duration_on_failure).await;
+                    sleep_duration_on_failure = (sleep_duration_on_failure * 2.0).min(MAX_SLEEP_DURATION);
+                },
+            };
+        }
+    }
+
+    pub(crate) async fn start_connection_loop(self, expires_at: Option<Instant>) {
+        let _guard = self.connection_guard.lock().await;
+
+        // List of awaiting requests
+        let mut response_notifiers: ExpirableMap<RequestId, oneshot::Sender<()>> = ExpirableMap::default();
+
+        let mut wsocket = match self
+            .attempt_to_establish_socket_connection(MAX_ATTEMPTS, SLEEP_DURATION)
+            .await
+        {
+            Ok(ws) => ws,
+            Err(e) => {
+                log::error!("Connection could not established for {}. Error {e}", self.node.uri);
+                return;
+            },
+        };
+
+        let mut keepalive_interval = Ticker::new(KEEPALIVE_DURATION);
+        let mut req_rx = self.controller_channel.rx.lock().await;
+
+        loop {
+            futures_util::select! {
+                _ = keepalive_interval.next().fuse() => {
+                    match self.handle_keepalive(&mut wsocket, &mut response_notifiers, expires_at).await {
+                        OuterAction::None => {},
+                        OuterAction::Continue => continue,
+                        OuterAction::Break => break,
+                        OuterAction::Return => return,
+                    }
+                }
+
+                request = req_rx.next().fuse() => {
+                    match self.handle_send_request(request, &mut wsocket, &mut response_notifiers).await {
+                        OuterAction::None => {},
+                        OuterAction::Continue => continue,
+                        OuterAction::Break => break,
+                        OuterAction::Return => return,
+                    }
+                }
+
+                message = wsocket.next().fuse() => {
+                    match self.handle_response(message, &mut response_notifiers).await {
+                        OuterAction::None => {},
+                        OuterAction::Continue => continue,
+                        OuterAction::Break => break,
+                        OuterAction::Return => return,
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn stop_connection_loop(&self) {
+        let mut tx = self.controller_channel.tx.lock().await;
+        tx.send(ControllerMessage::Close)
+            .await
+            .expect("receiver channel must be alive");
```

Comment:
also this one

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2058#discussion_r1499561248